# Практическая работа №2: Обработка выборочных данных. Нахождение точечных оценок параметров распределения

Выполнили студенты гр. 9383 Камзолов Никита и Крейсманн Кирилл. Вариант №3

## Цель работы
Получение практических навыков нахождения точечных статистических оценок параметров распределения.

## Основные теоретические требования

*Статистическое оценкой* $\Theta^*$ неизвестного параметра теоретического распределения $\Theta$ называется функция он наблюдаемых значений случайной величины:

$$\Theta^* = f(x_1,x_2,...,x_n)$$

Статистические оценки, определяемые одним числом, называются *точечными*.

Оценка называется *несмещенной*, если ее математическое ожидание равно оцениваемому параметру $\Theta$ при любом объеме выборке.
$$M(\Theta^*) = \Theta$$

Оценка $\Theta^*$ называется *эффективной*, если при заданном объеме выборки n она имеет наименьшую возможную дисперсию, т.е.:
$$D(\Theta^*)\rightarrow min$$

Оценка $\Theta^*$ называется *состоятельной* для параметра $\Theta$, если она сходится по вероятности к оцениваемому параметру при неограниченном возрастании объема выборке, т.е.:
$$\lim_{N\to \infty} P(|\Theta - \Theta^*|\leq\varepsilon) =1$$,
где $\varepsilon >0 $ - сколькоугодно малое положительное число.

Начальным эмпирическим моментом k-го порядка называется среднее значение k-х степенй элементов вариационного или интервального ряда.
$$\overline{M_k} = \frac{1}{N} \sum{n_j x_j^k}$$,
$n_j$ - абсолютная частота j-ой варианты, $x_j$ - j-ая варианта, N - объем выборки.

Выборочное среднее вычисляется по формуле:

$$\overline{x_в} = \overline{M_1} = \frac{1}{N} \sum n_j x_j$$

Центральным эмирическим моментом k-го порядка называется среднее значение k-х степеней разностей ($x_j - \overline{M_1}$) для вариационного или интервального ряда:
$$\overline{m_k} = \frac{1}{N} \sum{n_j(x_j - \overline{x_в})^k}$$

Выборочная дисперсия - центральный эмпирический момент второго порядка:
$$D_в = \overline{m_2} = \frac{1}{N} \sum{n_j (x_j - \overline{x_в})^2}$$

Исправленная оценка дисперсии:
$$s^2 = \frac{N}{N-1} D_в$$

Соответствующие среднеквадратические отклонения вычисляются как квадратный корень из соответствующих оценок дисперсии:
$$\sigma_в = \sqrt{D_в}; s = \sqrt{s^2}$$

Статистическая оценка асимметрии:
$$\overline{A_s} = \frac{\overline{m_3}}{s^3}$$

Статистическая оценка эксцесса:
$$\overline{E} = \frac{\overline{m_4}}{s^4} - 3$$

Для упрощения вычислений центральных эмпирических моментов вводятся условные эмпирические моменты и условные эмирические варианты.

Условными вариантами называются варианты, определенные равенством: 
$$u_j = \frac{x_j-C}{h}$$,
где C - условный ноль, значение которого выбирается равным значению варианты интервального ряда, являющейся средней или близкой к средней по значению в этом ряду ; h - длина интервала в интервальном ряду.

Условным эмпирическим моментом порядка k называется начальный момент порядка k, вычисленный для условных вариант:
$$\overline{M_k^*} = \frac{1}{N} \sum n_j (\frac{x_j-C}{h})^k = \frac{1}{N} \sum{n_j u_j^k}$$


В частности:

$$\overline{M_1^*} = \frac{1}{N} \sum n_j (\frac{x_j-C}{h}) = \frac{1}{N} \sum{n_j u_j}$$

Условные эмпирические моменты связаны с эмпирическими начальными и центральными моментами следующим образом:
$$\overline{x}_в = \overline{M}_1 = \overline{M}_1^* h + C$$
$$\overline{m}_2 = (\overline{M}_2^* - (\overline{M}_1^*)^2)h^2$$
$$\overline{m}_3 = (\overline{M}_3^* - 3\overline{M}_2^*\overline{M}_1^* + 2(\overline{M}_1^*)^3)h^3$$
$$\overline{m}_4 = (\overline{M}_4^* - 4\overline{M}_3^*\overline{M}_1^* + 6\overline{M}_2^*(\overline{M}_1^*)^2 - 3(\overline{M}_1^*)^4)h^4$$


Модой $M_o^*$ вариационного ряда называется варианта, которой соответствует наибольшая частота.

Для интервального ряда мода находится внутри интервала ($x_{i-1},x_i$), которому соответствует наибольшая частота и ее приближенное значение может быть вычислено по формуле:

$$M_o^* = x_{i-1}+h\frac{\tilde{m}_i - \tilde{m}_{i-1}}{(\tilde{m}_i - \tilde{m}_{i-1})+(\tilde{m}_i - \tilde{m}_{i+1})}$$
, где h - длина интервала; $\tilde{m}_i $ - относительная частота для i-го интервала.

Медианой $M_e^*$ называется варианта, соответствующая середине ранжированного ряда.

Если ранжированный ряд имеет нечетное (2m+1) количество элементов, то $M_e^* = x_{m+1}$.

Если ранжированный ряд имеет четное (2m) количество элементов, то $M_e^* = \frac{(x_m + x_{m+1})}{2}$

Для интервального ряда медиана вычисляется по формуле:
$$M_e^* = x_{i-1} + \frac{h}{\tilde{m}_i}(0.5 - \sum_{j=1}^{i-1}\frac{h}{\tilde{m}_j})$$


Выборочный коэффициент вариации $V^*$ характеризует разброс значений вариант относительно выборочного среднего:
$$V^* = \frac{\sigma_в}{\overline{x}_в}*100\%$$


















## Постановка задачи
Для заданных выборочных данных вычислить с использованием метода моментов и условных вариант точечные статистические оценки математического ожидания, дисперсии, среднеквадратичного отклонения, асимметрии и эксцесса исследуемой случайной величины. Полученные результаты содержательно проинтерпретировать.

## Выполнение работы

In [72]:
import pandas as pd
import numpy as np
import math

В лабораторной работе №1 был получен интервальный ряд.

|   | $[x_{i}, x_{i+1})$ | $\tilde{x}_{i}$ | $n_i$ | $p^*_i$ | $n_{i}^{\Sigma}$ | $p_{i}^{\Sigma}$ |
|:-:|:------------------:|:-------------:|:-----:|:-----:|:--------------:|:--------------:|
| 1 |    $[344, 379)$    |      361.5      |    11   | 0.102804  |        11       |       0.102804     |
| 2 |    $[379, 414)$    |      396.5     |  13  | 0.121495 |       24       |      0.224299     |
| 3 |    $[414, 449)$    |      431.5      |   35| 0.327103 |       59       |      0.551402     |
| 4 |    $[449, 484)$    |      466.5     |    20 | 0.186916 |        79      |      0.738318     |
| 5 |    $[484, 519)$    |      501.5     |   15  | 0.140187 |       94       |      0.878505     |
| 6 |    $[519, 554)$    |      536.5     |   9  | 0.0841121 |       103      |      0.962617     |
| 7 |    $[554, 585]$    |      569.5     |   4   | 0.0373832 |       107      |      1     |
|||$\sum$|107|1.0|-|-|-|

Количество интервалов k = 7; шаг h = 35; размер выборки N = 107.

In [73]:
k=7
h=35
N=107

#### Вычисление условных вариант для середин интервального ряда

In [129]:
left_bounds = [344,379,414,449,484,519,554]#левые границы интервалов
middle_points = [361.5, 396.5, 431.5, 466.5, 501.5, 536.5, 569.5] #середина i-го частичного интервала
absolute_freq = [11,13,35,20,15,9,4]#абсолютная частота i-го частичного интервала
relative_freq = [0.102804,0.121495,0.327103,0.186916,0.140187,0.0841121,0.0373832]#относительная частота i-го интервала
accum_relative_freq = [0.102804,0.224299,0.551402,0.738318,0.878505,0.962617,1]#накопленная частота для i-го интервала
C = middle_points[3]#ложный ноль

In [75]:
u_vars = [(i-C)/h for i in middle_points]#условные варианты
u_vars

[-3.0, -2.0, -1.0, 0.0, 1.0, 2.0, 2.942857142857143]

In [76]:
def add_sum(lst):#добавляет элемент в список и возвращает список
    res = lst.copy()
    res.append(sum(res))
    return res

In [87]:
#Рассчитаем данные для заполнения таблицы
n_u = [absolute_freq[i]*u_vars[i] for i in range(k)]
n_u2 = [absolute_freq[i]*u_vars[i]**2  for i in range(k)]
n_u3 = [absolute_freq[i]*u_vars[i]**3  for i in range(k)]
n_u4 = [absolute_freq[i]*u_vars[i]**4  for i in range(k)]
n_u1_4 = [absolute_freq[i]*(u_vars[i]+1)**4  for i in range(k)]
table = pd.DataFrame(
    {
        "$x_i$":add_sum(middle_points),#x_i
        "$n_i$":add_sum(absolute_freq),#абсолютные частоты
        "$u_i$":add_sum(u_vars),#услованые варианты
        "$n_i u_i$": add_sum(n_u),#n_i*u_i
        "$n_i u_i^2$":add_sum(n_u2),#n_i*u_i^2
        "$n_i u_i^3$":add_sum(n_u3),#n_i*u_i^3
        "$n_i u_i^4$":add_sum(n_u4),#n_i*u_i^4
        "$n_i (u_i+1)^4$":add_sum(n_u1_4),#n_i*(u_i+1)^4
    }
)#сформированная таблица
table.index = table.index + 1
# print(table.to_markdown())


|    |   $x_i$ |   $n_i$ |    $u_i$ |   $n_i u_i$ |   $n_i u_i^2$ |   $n_i u_i^3$ |   $n_i u_i^4$ |   $n_i (u_i+1)^4$ |
|---:|--------:|--------:|---------:|------------:|--------------:|--------------:|--------------:|------------------:|
|  $1$ |   361.5 |      $11$ | -3       |    -33      |       99      |      -297     |       891     |           $176$     |
|  $2$ |   396.5 |      $13$ | -2       |    -26      |       52      |      -104     |       208     |            $13$     |
|  $3$ |   431.5 |      $35$ | -1       |    -35      |       35      |       -35     |        35     |             $0$     |
|  $4$ |   466.5 |      $20$ |  0       |      0      |        0      |         0     |         0     |            $20$     |
|  $5$ |   501.5 |      $15$ |  1       |     15      |       15      |        15     |        15     |           $240$     |
|  $6$ |   536.5 |       $9$ |  2       |     18      |       36      |        72     |       144     |           $729$     |
|  $7$ |   569.5 |       $4$ |  2.94286 |     11.7714 |       34.6416 |       101.945 |       300.011 |           $966.728$ |
|  $\sum$ |  3263.5 |   $107$| - |    -49.2286 |      271.642  |      -247.055 |      1593.01  |          $2144.73$  |

Контроль вычислений:
$$\sum u_i^4 n_i + 4\sum u_i^3 n_i + 6\sum u_i^2 n_i + 4\sum u_i n_i + \sum n_i = 2144.73$$

#### Вычисление эмпирических моментов

In [96]:
#запишем суммы в переменные
sum_n_u = sum(n_u)
sum_n_u2 = sum(n_u2)
sum_n_u3 = sum(n_u3)
sum_n_u4 = sum(n_u4)

#Посчитаем условные эмпирические моменты
nu1 = sum_n_u/N
nu2 = sum_n_u2/N
nu3 = sum_n_u3/N
nu4 = sum_n_u4/N

#Посчитаем центральные эмпирические моменты
mu1 = nu1*h+C
mu2 = (nu2 - nu1**2)*h**2
mu3 = (nu3 - 3*nu1*nu2 +2*nu1**3)*h**3
mu4 = (nu4 - 4*nu3*nu1 + 6*nu2*nu1**2 -3*nu1**4)*h**4

table2 = pd.DataFrame(
    {
        "$\\nu_i^*$":[nu1,nu2,nu3,nu4],
        "$\mu_i^*$":[mu1,mu2,mu3,mu4]
    }
)#сформированная таблица
table2.index = table2.index + 1
print(table2.to_markdown())

|    |   $\nu_i^*$ |       $\mu_i^*$ |
|---:|------------:|----------------:|
|  1 |    -0.46008 |   450.397       |
|  2 |     2.53871 |  2850.62        |
|  3 |    -2.30892 | 42889.2         |
|  4 |    14.888   |     2.06015e+07 |


|    |   $\nu_i^*$ |       $\mu_i^*$ |
|---:|------------:|----------------:|
|  1 |    -0.46008 |   450.397       |
|  2 |     2.53871 |  2850.62        |
|  3 |    -2.30892 | 42889.2         |
|  4 |    14.888   |     2.06015e+07 |

#### Вычисление выборочной средней и дисперсии

In [97]:
#вычисление стандартным способом выборочного среднего
standart_mean = sum([absolute_freq[i]*middle_points[i] for i in range(k)])/N
#сравнение с вычисленным с помощью условных вариант
print(standart_mean, standart_mean == mu1)

450.39719626168227 True


In [100]:
#вычисление стандартным способом выборочной дисперсии
var = sum([absolute_freq[i]*(middle_points[i]-standart_mean)**2 for i in range(k)])/N
#сравнение с вычисленным с помощью условных вариант
print(var, var == mu2)

2850.6155996156867 True


#### Вычисление исправленной выборочной дисперсии и СКО.

In [103]:
s2 = var*N/(N-1)#исправленная выборочная дисперсия
s2

2877.5081996120607

In [104]:
s  = math.sqrt(s2)#исправленное СКО
s

53.64241045676509

In [105]:
sigma = math.sqrt(var)#смещенная СКО
sigma

53.391156567503636

## Тут сделать вывод какой-то

#### Вычисление статистической оценки коэффициентов асимметрии и эксцесса.

In [108]:
asym = mu3/s**3 # коэффициент асимметрии
asym

0.27785781776231133

In [109]:
exc = mu4/s**4 -3  # эксцесс
exc

-0.5119069355504227

Коэффициент асимметрии является положительной величиной, следовательно полигон частот и гистограмма распределения должны быть "скошены" больше вправо относительно самого высокого участка. Что подтверждается гистограммой из лабораторной работы №1.

Коэффициент эксцесса отрицателен. Следовательно эмпирическое распределение является более пологим относительно "эталонного" нормального распределения с параметрами $\alpha = \overline{x}_в, \sigma = \sigma_в$.

#### Вычисление моды, медианы и коэффициента вариации


In [118]:
index_mode = absolute_freq.index(max(absolute_freq))#индекс интервала с наибольшей частотой

mode = left_bounds[index_mode] + h*(relative_freq[index_mode] - relative_freq[index_mode-1])/\
    ((relative_freq[index_mode] - relative_freq[index_mode-1])+(relative_freq[index_mode]-relative_freq[index_mode+1]))
#приближенная мода, вычисленная для интервального ряда
mode

434.81082722422246

In [141]:
index_median = -1 #индекс интервала i, для которого накопленная частота >=0.5, а для интервала i-1 меньше 0.5
for i in range(len(accum_relative_freq)):
    if accum_relative_freq[i]>=0.5:
        index_median = i
        break

median =  left_bounds[index_median] +\
    (h/relative_freq[index_median])*(0.5-accum_relative_freq[index_median-1])
median #приближенная медиана, вычисленная для интервального ряда

443.4999893000064

In [143]:
variation =  sigma/standart_mean*100 #выборочный коэффициент вариации

variation

11.854238216989966

## Вывод

В ходе выполнения лабораторной работы получены практические навыки нахождения точечных статистических оценок параметров распределения.

Были вычислены условные варианты и условные эмпирические моменты для середин интервального ряда.

Вычислены выборочное среднее и дисперсия с помощью стандартных формул и с помощью условных вариант.

Вычисленна исправленная выборочная дисперсия и исправленное СКО.

Найдены статистические оценки коэффициентов асимметрии и эксцесса.

Для интервального ряда вычислены мода, медиана и коэффициент вариации.